### ISSUE TITLE
USB Init option?

### ISSUE THREAD

**Issue by xC0000005** [KevinOConnor/klipper#907]
I have Klipper running on the Malyan M200 V1/MPMini, and this particular board requires a pullup transistor on PB9 to be flipped on before USB init. Would the right way to do this be to add a config to the kconfig and then an #if based off of that?
People wanting to run on this are already going to have to make some changes, because the drivers require a minimum pulse length anyway, and it requires some custom startup assembly to cope with the oddity of the bootloader, so I've got no problems with it just being behind an #ifdef and telling people they need to change it.

**Comment by xC0000005** [comment]
I had marlin running on the mini a year and a half ago, and kept meaning to get around to porting Klipper to it. Once I saw you had an STM32F103 HAL, it was a few minutes of copying over my hard-learned hacks. Value line boards still use the D- pulldown. The Malyan M200V2/Pro, which is an STM32F070 board, for instance, has no pullup, and neither do any of the new L line boards from ST. 

Regarding the delay - 2 should be enough - that’s what I have to set in Marlin, but the behavior is absolutely identical - if I comment out the min pulse in marlin, my steppers buzz just like Klipper does. Maybe I’ve got something configured wrong, but the directions, endstops, heaters, fans, etc, all work. It’s just that the pulses seem to be firing *extremely quickly*. If I just change that (2) value in stepper.c to say, 4, that should be enough for me to test if that’s what’s really causing the problem.

I’ll add a section to the config, and one for the custom startup code (it’s a trivial change to the HAL layer). 

Thanks.

**Comment by KevinOConnor** [comment]
Any further update on this?

**Comment by xC0000005** [comment]
@clearchris is working on this with the mini. What I saw was that there were to issues probably related to my port - the Z axis moved, but moved erratically when not homing. The whole machine would freeze if I chose to restart the firmware (again, not likely related to the core). One problem I was sure would exist but couldn’t find was that the Malyan boards run at 48mhz and I suspect this was set up for 72mhz.

I’ll leave further updates to him for now.
